hathor/websocket/protocol.py

```diff
@@ -30,23 +42,230 @@ class HathorAdminWebsocketProtocol(WebSocketServerProtocol):
         self.log.debug('new message', payload=payload.hex() if isinstance(payload, bytes) else payload)
-        self.factory.handle_message(self, payload)
+        if isinstance(payload, bytes):
+            message = json_loadb(payload)
+        else:
+            message = json_loads(payload)
+
+        _type = message.get('type')
+
+        if _type == 'ping':
+            self._handle_ping(message)
+        elif _type == 'subscribe_address':
+            self.factory._handle_subscribe_address(self, message)
+        elif _type == 'unsubscribe_address':
+            self.factory._handle_unsubscribe_address(self, message)
+        elif _type == 'request:history:xpub':
+            self._open_history_xpub_streamer(message)
+        elif _type == 'request:history:manual':
+            self._open_history_manual_streamer(message)
+        elif _type == 'request:history:stop':
+            self._stop_streamer(message)
+
+    def _handle_ping(self, message: dict[Any, Any]) -> None:
+        """Handle ping message, should respond with a simple {"type": "pong"}"""
+        payload = json_dumpb({'type': 'pong'})
+        self.sendMessage(payload, False)
+
+    def _open_history_xpub_streamer(self, message: dict[Any, Any]) -> None:
+        """Handle request to stream transactions using an xpub."""
+        stream_id = message['id']
+
+        if self._history_streamer is not None:
+            self.sendMessage(json_dumpb({
+                'type': 'stream:history:error',
+                'id': stream_id,
+                'errmsg': 'Streaming is already opened.'
+            }))
+            return
+
+        xpub = message['xpub']
+        gap_limit = message.get('gap-limit', 20)
+        first_index = message.get('first-index', 0)
+        if gap_limit > self.MAX_GAP_LIMIT:
+            self.sendMessage(json_dumpb({
+                'type': 'stream:history:error',
+                'id': stream_id,
+                'errmsg': f'GAP limit is too big. Maximum: {self.MAX_GAP_LIMIT}'
+            }))
+            return
+
+        try:
+            address_iter = XPubAddressSequencer(xpub, first_index=first_index)
+        except InvalidXPub:
+            self.sendMessage(json_dumpb({
+                'type': 'stream:history:error',
+                'id': stream_id,
+                'errmsg': f'Invalid XPub: {xpub}'
+            }))
+            return
+
+        search = GAPLimitSearch(self.factory.manager, address_iter, gap_limit)
+        self._history_streamer = HistoryStreamer(protocol=self, stream_id=stream_id, search=search)
+        deferred = self._history_streamer.start()
+        deferred.addBoth(self._streamer_callback)
+        self.log.info('opening a websocket xpub streaming',
+                      stream_id=stream_id,
+                      xpub=xpub,
+                      gap_limit=gap_limit,
+                      first_index=first_index)
+
+    def _open_history_manual_streamer(self, message: dict[Any, Any]) -> None:
```

Comment:
That's right. It might be called multiples times for the same streamer. The first call would create it and the following calls would add addresses and optionally mark as last.

Reply:
Should we rename it, then?

Reply:
Sure! I just renamed it to `_handle_history_manual_streamer`.

Reply:
Thanks!